src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/dimensions/Dimension.java

```diff
@@ -56,18 +57,28 @@ public Dimension(WMS wms, String dimensionName, LayerInfo layerInfo, DimensionIn
      */
     public abstract List<Object> getDomainValues(Filter filter, boolean noDuplicates);
 
-
-    protected DomainSummary getDomainSummary(FeatureCollection features, String attribute, boolean includeCount) {
-        if (includeCount) {
-            Map<Aggregate, Object> aggregates = DimensionsUtils.getAggregates(attribute, features,
-                    Aggregate.MIN, Aggregate.MAX, Aggregate.COUNT);
-            return new DomainSummary(aggregates.get(Aggregate.MIN), aggregates.get(Aggregate.MAX), (Long)
-                    aggregates.get(Aggregate.COUNT));
-        } else {
-            Map<Aggregate, Object> aggregates = DimensionsUtils.getAggregates(attribute, features,
-                    Aggregate.MIN, Aggregate.MAX);
-            return new DomainSummary(aggregates.get(Aggregate.MIN), aggregates.get(Aggregate.MAX));
+    /**
+     * Returns the domain summary. If the count is lower than <code>expandLimit</code> then only the
+     * count will be returned, otherwise min and max will also be returned
+     * @param features
+     * @param attribute
+     * @param expandLimit
+     * @return
+     */
+    protected DomainSummary getDomainSummary(FeatureCollection features, String attribute, int expandLimit) {
+        // grab domain, but at most expandLimit + 1, to know if there are too many
+        if (expandLimit != 0) {
+            TreeSet uniqueValues = DimensionsUtils.getUniqueValues(features, attribute, expandLimit + 1);
+            if (uniqueValues.size() <= expandLimit || expandLimit < 0) {
+                return new DomainSummary(uniqueValues);
```

Comment:
In the case were the retrieved unique values are superior to the expand limit, will not be faster to compute the MIN and MAX than delegate everything again to the db ? Or let's be crazy compute those values on the fly when building the tree ?

Reply:
Computing all the unique values might take too much time and would basically ignore the settings configured by the admin on the limit and max limit. The getUniqueValues tries to get (expandLimit + 1) unique values, that is translated down into query, if the limits are properly configured, that will result in a fast query. If (expandLimit +1) are returned then there are too many, but we don't have them all, so delegating again to the db is the only option.

Reply:
Touché.